src/app/modules/submission/submission.service.ts

```diff
@@ -203,3 +207,58 @@ export const findSubmissionById = (
     return okAsync(submission)
   })
 }
+
+/**
+ * @param pendingSubmissionId the id of the pending submission to confirm
+ * @param session (optional) the mongoose transaction session to be used, if any
+ *
+ * @returns ok(submission document) if a submission document is successfully created
+ * @returns err(PendingSubmissionNotFoundError) if pending submission does not exist in the database
+ * @returns err(DatabaseError) if database errors occurs while copying the document over
+ */
+export const confirmPendingSubmission = (
```

Comment:
I think this is related to the other conversation so we can just have one thread for this. Context for writing this was the design we had before yesterday, where the responsibility of moving submissions was not coupled with updating the other payment-related data yet (e.g. receipt_url etc).

I had envisioned that this service would only be responsible for moving documents from `pendingSubmissions` to `submissions` and nothing else. I realized while writing the webhook handler that this specific part of the code was becoming pretty clunky and hard to read in the middle of all the other payments-related business logic. Therefore, I wanted to factor out this potentially more generic operation of moving documents from pending to complete, which could be used in places other than payments later on. 

Things have changed since then, obviously! Given the new design, I agree that we should have a single function that updates the payment document with the critical business logic in its own session. Now in terms of code design, how should we think about it? (Maybe it's a false dichotomy and you can point out if it is but) It feels like there is a balance between safety and maintainability here. Should the entire logic be encapsulated into one function, so that it's impossible for someone to use a subsection of the logic in an unintended way (safety)? Or, are we ok with still writing helpers (but maybe not exporting them, for readability)? As an illustration: the first option would look like 

```typescript
export const confirmPayment = (...args) => {
  await PendingSubmission.findById()
      .then((pendingSubmission) => { Submission.save(pendingSubmission) }) // of course, not as simple as this, and will probably look very hard to read
      .then((submission) => { Payment.update(paymentId, {submission.id}) } )
      // ...etc
```

whereas the second option _could_ look like
```typescript
const movePendingSubmissionToSubmission = (pendingSubmissionId) => {
  // identical to the confirmPendingSubmission function written in this PR
  await PendingSubmission.findById()
      .then((pendingSubmission) => { Submission.save(pendingSubmission) })
}

const updatePaymentWithSuccessMetadata = (...args) => {
  await Payment.update(...)
}

export const confirmPayment = (...args) => {
  // Much easier to read and more maintainable
  await movePendingSubmission(pendingSubmissionId)
    .then(updatePaymentWithSuccessMetadata)
```

Would be interested to hear what you/people think!

Reply:
Im buying option 2. Looks cleaner without any sacrifice to functionality or additional risks. Exporting the private functions for tests can be done through rewire too.

> Things have changed since then, obviously! 

Apologies for time travelling xD

Reply:
haha np! sounds good, will rewrite to accomodate this. In this case, the tradeoff here is that it might look a bit strange - the `movePendingSubmissionToSubmission` function will be in the payments service, despite not really touching the payments model at all. I think that's fine though, and we'll manage if we never need to generalize the function. thanks for the input!

Reply:
thanks for the elaborations! agree that the second option looks easier to understand